encoding/jsonc.ts

```diff
@@ -0,0 +1,372 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+import { assert } from "../_util/assert.ts";
+
+export interface ParseOptions {
+  /** Allow trailing commas at the end of arrays and objects. (default: `true`) */
+  allowTrailingComma?: boolean;
+}
+
+/**
+ * Converts a JSON with Comments (JSONC) string into an object.
+ * If a syntax error is found, throw a SyntaxError.
+ *
+ * @param text A valid JSONC string.
+ * @param options
+ * @param options.allowTrailingComma Allow trailing commas at the end of arrays and objects. (default: `true`)
+ *
+ * ```ts
+ * import * as JSONC from "https://deno.land/std@$STD_VERSION/encoding/jsonc.ts";
+ *
+ * JSONC.parse('{"foo": "bar", } // comment'); //=> { foo: "bar" }
+ * JSONC.parse('{"foo": "bar" } // comment', { allowTrailingComma: false }); //=> { foo: "bar" }
+ * ```
+ */
+export function parse(
+  text: string,
+  { allowTrailingComma = true }: ParseOptions = {},
+) {
+  if (new.target) {
+    throw new TypeError("parse is not a constructor");
+  }
+  return new JSONCParser(text, { allowTrailingComma }).parse() as JSONValue;
+}
+
+/** Valid types as a result of JSON parsing */
+export type JSONValue =
+  | { [key: string]: JSONValue }
+  | JSONValue[]
+  | string
+  | number
+  | boolean
+  | null;
+
+enum tokenType {
+  beginObject,
+  endObject,
+  beginArray,
+  endArray,
+  nameSeparator,
+  valueSeparator,
+  nullOrTrueOrFalseOrNumber,
+  string,
+}
+
+type tokenized = {
+  type: Exclude<
+    tokenType,
+    tokenType.string | tokenType.nullOrTrueOrFalseOrNumber
+  >;
+  sourceText?: undefined;
+  position: number;
+} | {
+  type: tokenType.string;
+  sourceText: string;
+  position: number;
+} | {
+  type: tokenType.nullOrTrueOrFalseOrNumber;
+  sourceText: string;
+  position: number;
+};
+
+const originalJSONParse = globalThis.JSON.parse;
+
+// First tokenize and then parse the token.
+class JSONCParser {
+  readonly #whitespace = new Set(" \t\r\n");
+  readonly #numberEndToken = new Set([..."[]{}:,/", ...this.#whitespace]);
+  #text: string;
+  #length: number;
+  #tokenized: Generator<tokenized, void>;
+  #options: ParseOptions;
+  constructor(text: string, options: ParseOptions) {
+    this.#text = `${text}`;
```

Comment:
How about just assigning `text` to `#text`?
```suggestion
    this.#text = text;
```

Reply:
Other JSONC parsers and native JSON.parse cast the type to a string, so I wanted to match the behavior.
https://github.com/fabiospampinato/jsonc-simple-parser/blob/370af65f2ea6c40d5faa060e315e4d289a86f7df/src/parse.ts#L12-L15

If implicit type conversion is not preferred here, it can also be written as `if (typeof text!=="string") { throw new Error(); }`. Is this better?

Reply:
Ah, ok. This looks fine now